### PR TITLE
index hints for traversals

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

* Support indexHint AQL option for traversals.
  The structure of traversal index hints is as follows:
  
      { collection: { direction: { level: indexes } } }
 
  with the following meanings:
  - collection: the name of an edge collection for which the index hint should
    be applied. Proper capitalization of the collection name is important.
  - direction:  the direction for which to apply the index hint. Valid values
    are `inbound` and `outbound`, in lowercase.
  - level: the level/depth for which the index should be applied. Valid level 
    values are the string `base` and any stringified numeric values greater
    or equal to zero.

  Example:

      FOR v, e, p IN 1..5 OUTBOUND startVertex edgeCollection
      OPTIONS { 
        indexHint: {
          "edgeCollection": {
            "outbound": {
              "base": ["edge"],
              "1": "myIndex1",
              "2": ["myIndex2", "myIndex1"],
              "3": "myIndex3",
            }
          }
        }
      }
      FILTER p.edges[1].foo == 'bar' && 
             p.edges[2].foo == 'bar' && 
             p.edges[2].baz == 'qux'

  Because the collection names and levels/depths are used a object keys here,
  it is useful to enclose them in quotes. Omitting the quotes around them may
  result in parse errors for the query.

  Note that index hints for levels other than `base` will only be considered
  if the traversal actually uses a specific filter condition for the specified
  level. In the above example, this is true for levels 1 and 2, but not for
  level 3. Consequently, the index hint for level 3 will be ignored here.

  Currently index hints cannot be forced to be used for traversals, i.e. the
  `forceIndexHint` option is currently not supported for traversals.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.12.0: -
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,51 @@
...
+  Example:
+
+      FOR v, e, p IN 1..5 OUTBOUND startVertex edgeCollection
+      OPTIONS { 
+        indexHint: {
+          "edgeCollection": {
+            "outbound": {
+              "base": ["edge"],
+              "1": "myIndex1",
+              "2": ["myIndex2", "myIndex1"],
```
Comment (maierlars) [inline CHANGELOG:27]: what does it mean if one specifies an array?

File: CHANGELOG [hunk 2]
```diff
@@ -1,6 +1,51 @@
...
+  it is useful to enclose them in quotes. Omitting the quotes around them may
+  result in parse errors for the query.
+
+  Note that index hints for levels other than `base` will only be considered
+  if the traversal actually uses a specific filter condition for the specified
+  level. In the above example, this is true for levels 1 and 2, but not for
+  level 3. Consequently, the index hint for level 3 will be ignored here.
+
+  Currently index hints cannot be forced to be used for traversals, i.e. the
+  `forceIndexHint` option is currently not supported for traversals.
```
Comment (maierlars) [inline CHANGELOG:47]: how much work is that? have you looked into it?

File: CHANGELOG (hunk 1 above)
Comment (jsteemann) [reply CHANGELOG:27]: It means the same as when specifying an array index hint for a collection `FOR` loop.
For example, one can already use something like
```
FOR doc IN collection OPTIONS { indexHint: ["index1", "index2"] }
...
```
even without this change. The candidate indexes for an index hint can be specified as a single string (one index, identified by its string name) or as an array of strings (multiple indexes, identified by their string names).
This is the same for traversal index hints: the candidate indexes for each level can be specified as a single string or as an array of strings.

File: CHANGELOG (hunk 2 above)
Comment (jsteemann) [reply CHANGELOG:47]: I looked into this. It is a bit unclear to me what `forceIndexHint: true` should actually mean for a traversal index hint.
A traversal index hint can consist of multiple sub-index hints, of which any can be satisfied (i.e. used) or not satisfied. It is unclear to me what `forceIndexHint: true` actually means if one some sub-index hints are picked up, or if there are more index hints given than what are actually used (e.g. if user specifies an index hint for a level which is not really relevant).

File: arangod/Aql/ExecutionPlan.cpp [hunk 3]
```diff
@@ -406,10 +418,30 @@ std::unique_ptr<graph::BaseOptions> createShortestPathOptions(
         TRI_ASSERT(value->isConstant());
 
         if (name == "weightAttribute" && value->isStringValue()) {
-          options->setWeightAttribute(
-              std::string(value->getStringValue(), value->getStringLength()));
+          options->setWeightAttribute(value->getString());
         } else if (name == "defaultWeight" && value->isNumericValue()) {
           options->setDefaultWeight(value->getDoubleValue());
+        } else if (name == StaticStrings::IndexHintOptionForce) {
+#ifndef ENABLE_FORCED_INDEX_HINTS_FOR_GRAPH_OPERATIONS
```
Comment (maierlars) [inline arangod/Aql/ExecutionPlan.cpp:425]: Is this macro completely implemented or shall we just drop it? Or do you just want to keep it (which is fine by me)

**Review by maierlars — APPROVED** [review]

**Pull request merged**